core/data/src/main/java/com/google/samples/apps/nowinandroid/core/data/util/ConnectivityManagerNetworkMonitor.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.samples.apps.nowinandroid.core.data.util
+
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.ConnectivityManager.NetworkCallback
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkRequest.Builder
+import android.os.Build.VERSION
+import android.os.Build.VERSION_CODES
+import dagger.hilt.android.qualifiers.ApplicationContext
+import javax.inject.Inject
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+
+class ConnectivityManagerNetworkMonitor @Inject constructor(
+    @ApplicationContext private val context: Context
+) : NetworkMonitor {
+    override val isOnline: Flow<Boolean> = callbackFlow {
+        val callback = object : NetworkCallback() {
+            override fun onAvailable(network: Network) {
+                channel.trySend(true)
+            }
+
+            override fun onLost(network: Network) {
+                channel.trySend(false)
+            }
+        }
+
+        val connectivityManager = context.getSystemService(
+            Context.CONNECTIVITY_SERVICE
+        ) as? ConnectivityManager
+
+        connectivityManager?.registerNetworkCallback(
+            Builder()
+                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+                .build(),
+            callback
+        )
+
+        channel.trySend(connectivityManager.isCurrentlyConnected())
+
+        awaitClose {
+            connectivityManager?.unregisterNetworkCallback(callback)
+        }
+    }
```

Comment:
Nit: We could consider adding `.conflate()` to the end here, since I would guess we should only care about the most recent value (and therefore `trySend` should never fail)

Reply:
Done, why `conflate()` over `distinctUntilChanged()`?

Reply:
I think `distinctUntilChanged()` doesn't change the backpressure like `conflate` does. If we had a really slow collector, and were emitting `true`, `false`, `true`, `false`, etc., then `distinctUntilChanged` wouldn't have any effect.